src/index.ts

```diff
@@ -194,6 +194,35 @@ export class SgidClient {
     }
     return result
   }
+
+  /**
+   * Generates the code verifier (random bytes encoded in url safe base 64) to be used in the OAuth 2.0 PKCE flow
+   * @param length The length of the code verifier to generate (Defaults to 43 if not provided)
+   * @returns The generated code verifier
+   */
+  codeVerifier(length = 43): string {
+    if (length < 43 || length > 128) {
+      // eslint-disable-next-line typesafe/no-throw-sync-func
+      throw new Error(
+        `The code verifier should have a minimum length of 43 and a maximum length of 128. Length of ${length} was provided`,
+      )
+    }
+
+    // 96 bytes results in a 128 long base64 string
+    const codeVerifier = generators.codeVerifier(96)
+
+    // This works because a prefix of a random string is still random
+    return codeVerifier.slice(0, length)
+  }
+
+  /**
+   * Calculates the S256 PKCE code challenge for a provided code verifier
+   * @param codeVerifier The code verifier to calculate the S256 code challenge for
+   * @returns The calculated code challenge
+   */
+  codeChallenge(codeVerifier: string): string {
+    return generators.codeChallenge(codeVerifier)
+  }
```

Comment:
I'd personally merge these two functions into one as they have to be called in pair.  
My thoughts are 
generatePkceChallenge(length?:number) => 
```
{
    code_verifier: 'u1ta-MQ0e7TcpHjgz33M2DcBnOQu~aMGxuiZt0QMD1C',
    code_challenge: 'CUZX5qE8Wvye6kS_SasIsa8MMxacJftmWdsIA_iKp3I'
}
```

Thoughts? @kwajiehao @mantariksh @kschiew

Reply:
This discussion will determine the implementation for all other SDKs too.

Reply:
Good idea, I think it might help to reduce confusion on the developer side by exposing less functions as well

Reply:
From the users' perspective, I think it's a good idea to have a `generatePkcePair` function that generates both the challenge and verifier.

Personally, I agree with what @mantariksh said in the implementation sketch, which is to provide good defaults but also flexibility. We can make it clear in the documentation that using our `generatePkcePair` is strongly recommended, but we should also expose the two functions in the event they have some weird use case(s).

Referencing the two implementations mentioned by @PrawiraGenestonlia, this seems to be what they did as well.

Reply:
yup got it, that makes a lot of sense - thanks!